cedar-policy-core/src/entities/json/schema_types.rs

```diff
@@ -319,6 +312,40 @@ pub struct HeterogeneousSetError {
     ty2: Box<SchemaType>,
 }
 
+/// Trying to compute the [`SchemaType`], but the value or expression
+/// contains an [`Unknown`] that has insufficient type information
+/// associated in order to compute the `SchemaType`
+#[derive(Debug, Diagnostic, Error)]
+pub struct UnknownInsufficientTypeInfoError {
+    /// `Unknown` which has insufficient type information
+    unknown: Option<Unknown>,
+}
+
+impl std::fmt::Display for UnknownInsufficientTypeInfoError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "cannot compute type because of insufficient type information for"
+        )?;
+        match &self.unknown {
+            Some(u) => write!(f, "`{u}`"),
+            None => write!(f, "unknown"),
+        }
+    }
+}
+
+/// Trying to compute the [`SchemaType`] of a nontrivial residual (i.e., a
+/// residual which is not just a single `Unknown`). For now, we do not
+/// attempt to compute the [`SchemaType`] in these cases, and just return
+/// this error.
+#[derive(Debug, Diagnostic, Error)]
+#[error("cannot compute type of nontrivial residual `{residual}`")]
+pub struct NontrivialResidualError {
+    /// Nontrivial residual which we were trying to compute the
+    /// [`SchemaType`] of
+    residual: Box<Expr>,
+}
```

Comment:
bonus points for a `Diagnostic` implementation to take `labels()` and `source_code()` from the expr's `Loc`.  There are other errors you can copy-paste this implementation from (or make a macro)

Reply:
done